suite-native/device/src/middlewares/deviceConnectionMiddleware.ts

```diff
@@ -0,0 +1,87 @@
+import {
+    Dispatch,
+    ListenerEffectAPI,
+    UnknownAction,
+    createListenerMiddleware,
+} from '@reduxjs/toolkit';
+
+import {
+    deviceConnectThunks,
+    selectIsDeviceConnectedAndAuthorized,
+    selectIsDeviceInitialized,
+    selectIsDeviceUsingPassphrase,
+} from '@suite-common/wallet-core';
+import {
+    AuthorizeDeviceStackRoutes,
+    RootStackRoutes,
+    navigationContainerRef,
+} from '@suite-native/navigation';
+import { selectIsCoinEnablingInitFinished } from '@suite-native/settings';
+
+import { NativeDeviceRootState, selectIsDeviceCompromised } from '../selectors';
+
+export const deviceConnectionMiddleware = createListenerMiddleware<NativeDeviceRootState>();
+
+const connectDevice = ({ isCoinEnablingInitFinished }: { isCoinEnablingInitFinished: boolean }) => {
+    // If coin enabling is not finished, it takes priority over connecting screen
+    if (isCoinEnablingInitFinished) {
+        navigationContainerRef.navigate(RootStackRoutes.AuthorizeDeviceStack, {
+            screen: AuthorizeDeviceStackRoutes.ConnectingDevice,
+        });
+    } else {
+        navigationContainerRef.navigate(RootStackRoutes.CoinEnablingInit);
+    }
+};
+
+export const startDeviceConnectionListening = () => {
+    deviceConnectionMiddleware.startListening({
+        predicate: (action, currentState) =>
+            deviceConnectThunks.fulfilled.match(action) &&
+            // TODO this should dissappear after we merge device onboarding redirect here as well
+            // https://github.com/trezor/trezor-suite/issues/20157
+            // If device is not initialized and is compromised, we display the modal (reason why this condition is here) and then want to redirect to uninitialized device landing.
+            (selectIsDeviceInitialized(currentState) || selectIsDeviceCompromised(currentState)),
+        effect: (
+            _action,
+            { getState }: ListenerEffectAPI<NativeDeviceRootState, Dispatch<UnknownAction>>,
+        ) => {
+            // Passphrase protected devices are only connected through passphrase form
+            // The passphrase flow handles connection differently and redirect to connecting screen is not wanted.
+            const isDeviceUsingPassphrase = selectIsDeviceUsingPassphrase(getState());
+
+            // If device is authorized already (usually in case of remembered device which has already been authorized)
+            const isDeviceConnectedAndAuthorized = selectIsDeviceConnectedAndAuthorized(getState());
+
+            if (isDeviceUsingPassphrase || isDeviceConnectedAndAuthorized) return;
```

Comment:
I'm not sure about the position of this condition, wouldn't be better to have it below shouldNavigateToDeviceCompromisedModal to not skip it?